### PR TITLE
[REEF-324]:Propagate clr tcpprovider bindings to java

### PR DESCRIPTION
**Pull request by beysims**
This addressed the issue by passing specific tcp bindings from clr to
java side

JIRA: [REEF-324](https://issues.apache.org/jira/browse/REEF-324)

This closes #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Wake.Remote.Parameters;
+
+namespace Org.Apache.REEF.Client.Common
+{
+    public class ClrClient2JavaClientCuratedParameters
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs]: Does this need to be `public`? If so, please document.

File: lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿using Org.Apache.REEF.Tang.Annotations;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs]: Please add license headers.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java [hunk 3]
```diff
@@ -54,18 +57,21 @@
...
+  private final Set<ConfigurationProvider> configurationProviders;
 
   @Inject
   YarnJobSubmissionClient(final JobUploader uploader,
                           final YarnConfiguration yarnConfiguration,
                           final ConfigurationSerializer configurationSerializer,
                           final REEFFileNames fileNames,
-                          final ClasspathProvider classpath) {
+                          final ClasspathProvider classpath,
+                          final @Parameter(DriverConfigurationProviders.class) Set<ConfigurationProvider> configurationProviders) {
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: Why do we need these here? Wouldn't the config be merged in in `REEFImplementation` anyhow?

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/ActiveContextBridgeFactory.java [hunk 4]
```diff
@@ -35,7 +35,7 @@
 @ThreadSafe
 @Private
 public final class ActiveContextBridgeFactory {
```
Comment (markusweimer) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/ActiveContextBridgeFactory.java]: The changes in this file are now redundant with the ones merged as REEF-328, right? If so, please remove them.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/client/REEFImplementation.java [hunk 5]
```diff
@@ -96,11 +96,11 @@ public void submit(final Configuration driverConf) {
       final JobSubmissionEvent submissionMessage;
       try {
         if (this.clientWireUp.isClientPresent()) {
-          submissionMessage = this.jobSubmissionHelper.getJobSubmissionBuilder(driverConf)
+          submissionMessage = this.jobSubmissionHelper.getJobSubmissionBuilder(driverConfiguration)
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/client/REEFImplementation.java]: This fixes a genuine bug in `REEFImplementation`, right? We should call this out in the commit message.

File: lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameServerImpl.java [hunk 6]
```diff
@@ -87,7 +89,6 @@ public NameServerImpl(
   @Deprecated
   public NameServerImpl(final int port, final IdentifierFactory factory) {
```
Comment (markusweimer) [inline lang/java/reef-io/src/main/java/org/apache/reef/io/network/naming/NameServerImpl.java]: What are the changes here and why are they needed?

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/DriverConfigurationProvider.java [hunk 7]
```diff
@@ -35,12 +36,15 @@
 
   private final int maxEvaluators;
   private final double jvmHeapSlack;
+  private final Set<ConfigurationProvider> configurationProviders;
 
   @Inject
   DriverConfigurationProvider(final @Parameter(MaxNumberOfEvaluators.class) int maxEvaluators,
-                              final @Parameter(JVMHeapSlack.class) double jvmHeapSlack) {
+                              final @Parameter(JVMHeapSlack.class) double jvmHeapSlack,
+                              final @Parameter(DriverConfigurationProviders.class) Set<ConfigurationProvider> configurationProviders)  {
```
Comment (markusweimer) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/DriverConfigurationProvider.java]: Why do we need this in addition to the merge that happens in `REEFImplementation`?

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java [hunk 8]
```diff
@@ -80,6 +83,19 @@
    * the Driver (and the Evaluators) also use it.
    */
   public static final OptionalImpl<TcpPortProvider> TCP_PORT_PROVIDER = new OptionalImpl<>();
+  /**
```
Comment (markusweimer) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java]: Are these needed for the client? I though the issue only exists on the Driver / the Evaluators?

**Comment by markusweimer** [comment]
I did a pass. My main concern is that I don't understand the need for many of the changes. Shouldn't this "only" change the clients to accept a couple of ints which they add to the `DriverConfiguration`?

Also, the change contains conflicts with current master, please resolve.

File: lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs (hunk 2 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs]: Added license headers.

File: lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs (hunk 1 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/Common/ClrClient2JavaClientCuratedParameters.cs]: Changed to internal.

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/ActiveContextBridgeFactory.java (hunk 4 above)
Comment (beysims) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/ActiveContextBridgeFactory.java]: Removed

**Comment by markusweimer** [comment]
@beysims this is currently not ready to be merged, right? Can you please close the PR?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalClient.java [hunk 9]
```diff
@@ -6,9 +6,9 @@
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
- *
- *   http://www.apache.org/licenses/LICENSE-2.0
- *
+ * <p/>
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalClient.java]: Please undo the change to the license header. See [REEF-315](https://issues.apache.org/jira/browse/REEF-315)

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java (hunk 3 above)
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnJobSubmissionClient.java]: This still confuses me.

File: lang/java/reef-common/src/main/java/org/apache/reef/io/TcpPortConfigurationProvider.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/io/TcpPortConfigurationProvider.java]: Please use the non-javadoc form for comments and remove the `<p>` tags. See [REEF-315](https://issues.apache.org/jira/browse/REEF-315)

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/DriverConfigurationProvider.java (hunk 7 above)
Comment (markusweimer) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/DriverConfigurationProvider.java]: The question remains.

**Comment by markusweimer** [comment]
@beysims I did a pass. Mostly auto-format induced licensing issues. Also, some of my earlier comments weren't answered. I left a second comment on those where I am still confused.

**Comment by beysims** [comment]
Pushed a new iteration addressing feedback

**Comment by markusweimer** [comment]
I'll do a pass.

**Comment by markusweimer** [comment]
merging it now.
